send reminder emails for draft versions of deposits

jmartin-sul opened the pull request:
## Why was this change made?

implement #589.  from the [spec](https://docs.google.com/document/d/1TZVILUrORI25kRZyUZ70vMQYo90Ngp_tqCI7BdU_VyQ/edit#heading=h.lj2h48ox03kd):
> "How much time has elapsed before the reminder is issued?  How many reminders are sent before they cease?"
>
> "My inclination is to go with two weeks, and then a month, and then monthly thereafter.  I wouldn't stop as long as the SUNet is valid, but it would be nice if they had an option to request that they stop.  That might be out of scope right now."

This PR as is has a bit of a delta from that description:  I'd originally interpreted as notify at two weeks, then every four weeks after that... but upon re-reading, I think the intent is notify at two weeks, again at four weeks, then every four weeks after that?  If so, it should be easy enough to make that change in this PR or in a follow on (happy to do either).

## How was this change tested?

unit tests.  some manual testing locally on rails console and using the mailer preview, e.g.:
<img width="1414" alt="Screen Shot 2020-12-10 at 9 38 02 AM" src="https://user-images.githubusercontent.com/7741604/101809407-60667980-3acc-11eb-8b3b-f1e08181651b.png">


## Which documentation and/or configurations were updated?

Added the `whenever` gem and an initial daily cron job to run the new functionality.

jcoyne commented on app/mailers/works_mailer.rb:
```diff
@@ -15,6 +15,15 @@ def deposited_email
+    @work = params[:work]
+    @user = @work.depositor
+
+    subject = "Reminder: New version of a deposit to the #{@work.collection.name} collection in the SDR is in progress"
```
can you use `@work.collection_name` here so that this object doesn't need to depend on a deep graph of dependencies?

jmartin-sul replied:
will do...

jcoyne commented on app/models/work.rb:
```diff
@@ -88,6 +88,30 @@ class Work < ApplicationRecord
   end
 
+  sig { params(first_interval: ActiveSupport::Duration, subsequent_interval: ActiveSupport::Duration).void }
+  def self.send_first_draft_reminders(first_interval: nil, subsequent_interval: nil)
```
Can you move all of this to a service object (or background job)?  A models responsibility is about persistence, and this code is about sending mails.

jmartin-sul replied:
ah yeah, will do, had intended to do that and then forgot to refactor...

jcoyne reviewed (changes requested):
I believe this also needs to handle version_drafts too.

jmartin-sul commented:
@jcoyne i think i've addressed all the suggestions.  had to re-gen `.rbi` files to get the type-checking to pass after i extracted `send_first_draft_reminders` and `send_first_draft_reminder?` to a service class.  sorry for the churn, stuck it in a commit of its own.

jcoyne commented on app/services/work_reminder_generator.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+  def self.send_first_draft_reminder?(work, first_interval, subsequent_interval)
+    today = DateTime.now.utc.beginning_of_day
+    first_reminder_day = work.created_at.utc.beginning_of_day.to_datetime + first_interval
```
Should this be `updated_at`?  Since we're looking at a time since they last touched this item.  Is there a reason we can't run this query in the database (e.g. show me the works that are in a draft state and haven't been touched for 2 weeks?)

jmartin-sul replied:
Hmm... so I was initially using `created_at` when I thought this only applied to the `:first_draft` state.  But additionally, I was also under the impression that the reminder interval should be based on when the work went _into_ the draft state, and not when it was last edited.

But looking back at the issue and the spec it links, that now seems ambiguous to me, and I don't think my assumption there was clearly founded.  @jcoyne or @amyehodge, could you clarify whether the desired behavior is for the notification interval to be based on how long the work has been in draft state, vs basing it on the last edit?  Even if it's the former (how long it's been in draft state), I think I could join to `events` to get that info, instead of looking at ~~`created_at`~~`updated_at`.  But, now that I'm not just looking at things in `:first_draft`, using `created_at` does clearly feel wrong no matter what.

@jcoyne, I had briefly considered limiting by date in the query, but I didn't have a strong preference about that when mulling it over.  After getting into the implementation, it seemed a bit easier to do the date math that would let me get everything from an entire given day in the Ruby code instead of in the query.  But it would also be easy enough to filter in the query with a cutoff that's one day later than it should be, and then to keep the same filtering in the Ruby code.  Or, I could just do the date math in SQL afterall 🙂   Do you have a strong preference on this front?

jmartin-sul replied:
oh actually, i suppose i could also just calculate the cutoff date in ruby code and pass that into the query.  so yeah, easy enough to do some date filtering in the query, if you'd prefer.

jcoyne replied:
It's not a big deal now, but I can imagine there might be thousands of objects in a draft states, so a sql query would prevent us from having to "fix" this down the road.

jmartin-sul replied:
yeah, i didn't have a great sense of what the expected scale would be, so i wasn't sure how important it'd be to optimize that part of things.  will fix the query.

jcoyne commented on app/services/work_reminder_generator.rb:
```diff
@@ -0,0 +1,40 @@
+    first_interval ||= Settings.notifications.first_draft_reminder.first_interval.days
+    subsequent_interval ||= Settings.notifications.first_draft_reminder.subsequent_interval.days
+
+    Work.with_state(:first_draft, :version_draft).find_each do |work|
```
Sorry, It looks like `version_draft` should be part of #592

jmartin-sul replied:
ah ok, will back that change out, no worries

jmartin-sul commented:
After playing with it a bit more and looking up the applicable date functions, I ended up pushing all the selection logic into the query.  For whatever reason, I was thinking it'd be uglier, with more casting or formatting, but the SQL version is actually a lot more concise than the ruby version was.  To support that shift, I indexed `works.created_at` and `works.updated_at`.  I left that change in its own commit (only required a very minor test adjustment, since passing the intervals in as `Integer`s started to seem more sensible than using `ActiveSupport::Duration`s, since the SQL wants ints anyway).

jmartin-sul commented:
I also tacked on a commit that switched back to using `created_at`, since that was easy to do now that `:first_draft` is the only state in consideration.  It seemed to make sense given the length of the notification interval.  But it's super easy to revert back to using `updated_at`, should that be the way to go.

jcoyne commented:
@jmartin-sul You have a small conflict in Gemfile.lock to resolve, then please merge.
